### PR TITLE
Keep the linear constraint projection scaling constant

### PR DESCRIPTION
**Pull request by YigitElma**
Keep the linear constraint projection scaling constant throughtout the optimization

### CODE REVIEW

File: desc/optimize/_constraint_wrappers.py [hunk 1]
```diff
@@ -351,7 +338,7 @@ def grad(self, x_reduced, constants=None):
         """
         x = self.recover(x_reduced)
         df = self._objective.grad(x, constants)
-        return df[self._unfixed_idx] @ (self._Z * self._D[self._unfixed_idx, None])
+        return df[self._unfixed_idx] @ self._ZA
```
Comment (dpanici) [inline desc/optimize/_constraint_wrappers.py]: I believe we want this to retain the scaling matrix D (`_ZA` is `self._Z * self._D[self._unfixed_idx, None]` but then its columns are scaled to be unit norm, so it is not a drop-in replacement)

File: desc/optimize/_constraint_wrappers.py [hunk 2]
```diff
@@ -374,7 +361,7 @@ def hess(self, x_reduced, constants=None):
         return (
             (self._Z.T * (1 / self._D)[None, self._unfixed_idx])
             @ df[self._unfixed_idx, :][:, self._unfixed_idx]
-            @ (self._Z * self._D[self._unfixed_idx, None])
+            @ self._ZA
```
Comment (dpanici) [inline desc/optimize/_constraint_wrappers.py]: same here

File: desc/optimize/_constraint_wrappers.py [hunk 3]
```diff
@@ -491,7 +478,7 @@ def jvp_unscaled(self, v, x_reduced, constants=None):
     def _vjp(self, v, x_reduced, constants=None, op="vjp_scaled"):
         x = self.recover(x_reduced)
         df = getattr(self._objective, op)(v, x, constants)
-        return df[self._unfixed_idx] @ (self._Z * self._D[self._unfixed_idx, None])
+        return df[self._unfixed_idx] @ self._ZA
```
Comment (dpanici) [inline desc/optimize/_constraint_wrappers.py]: same here

**Review by dpanici — CHANGES_REQUESTED** [review]
undo the _ZA change, maybe instead make another self attribute for the un-normalized ZA and use that

**Review by dpanici — DISMISSED** [review]

File: desc/objectives/utils.py [hunk 4]
```diff
@@ -120,6 +122,7 @@ def factorize_linear_constraints(objective, constraint, x_scale="auto"):  # noqa
 
     # compute x_scale if not provided
     # Note: this x_scale is not the same as the x_scale as in solve_options["x_scale"]
+    # but the one given as solve_options["linear_constraint_options"]["x_scale"]
```
Comment (rahulgaur104) [inline desc/objectives/utils.py:125]: Really! Should we add a comment stating this somewhere?

File: desc/optimize/_constraint_wrappers.py [hunk 5]
```diff
@@ -232,19 +236,21 @@ def update_constraint_target(self, eq_new):
...
-        # pinv(A) = Ainv, ADinv = pinv(A @ D) = Dinv @ Ainv, Dinv = 1 / D
-        self._ADinv = (1 / self._D)[unfixed_idx, None] * self._Ainv
-        # we also need to update the nullspace Z of AD in a similar way
-        # A @ ZA = 0 -> (A @ D) @ ((1 / D) @ ZA) = 0 -> Z = (1 / D) @ ZA
-        # where ZA is the nullspace of A, and Z is the nullspace of AD
-        self._Z = (1 / self._D)[self._unfixed_idx, None] * self._ZA
-        # we also normalize Z to make each column have unit norm
-        self._Z = self._Z / jnp.linalg.norm(self._Z, axis=0)
+        # if user specified x_scale, don't dynamically change it
+        if self._x_scale == "auto":
```
Comment (rahulgaur104) [inline desc/optimize/_constraint_wrappers.py:240]: From the comment it sounds like when the self._x_scale="auto" means user-specified x_scale?
But from the if condition it looks like self._xscale="auto" means it's automatically chosen.

**Review by rahulgaur104 — DISMISSED** [review]
Some more clarification would be nice.

File: desc/objectives/utils.py (hunk 4 above)
Comment (rahulgaur104) [reply desc/objectives/utils.py:125]: I guess you do add it in the docstring later.

File: desc/optimize/_constraint_wrappers.py (hunk 5 above)
Comment (YigitElma) [reply desc/optimize/_constraint_wrappers.py:240]: Yeah, if it is `auto` then we choose it by values of the parameters. The `solve_options["linear_constraint_options"]["x_scale"]` was added recently, we forgot to update this part. The code was using the given x_scale only for 1 optimization iteration (for eq solve this is not a problem), then moving to computing it automatically!

File: desc/objectives/utils.py (hunk 4 above)
Comment (YigitElma) [reply desc/objectives/utils.py:125]: @ddudt added this option in #1697, this is briefly mentioned here https://github.com/PlasmaControl/DESC/blob/1d4806d280d11722839a73aee74f0888d07eeca1/desc/optimize/optimizer.py#L144

And with the change in `LinearConstraintProjection`s docs, it is complemented, but yes, this is a bit hard to find.

**Review by dpanici — APPROVED** [review]

**Review by ddudt — APPROVED** [review]
